aten/src/ATen/core/autocast/autocast_mode.cpp

```diff
@@ -0,0 +1,60 @@
+#include <ATen/ATen.h>
+#include <ATen/core/autocast/autocast_mode.h>
+#include <c10/core/impl/LocalTensorTypeSet.h>
+
+#include <stdexcept>
+#include <memory>
+#include <unordered_map>
+
+namespace at {
+namespace autocast {
+
+/// thread_local is a feature that is not enabled by Caffe2 mobile
+/// build (e.g. iOS). Therefore, we only provide `at::AutocastMode`
+/// when we are not in mobile build or when FEATURE_TORCH_MOBILE
+/// is on.
+#if !defined(C10_MOBILE) || defined(FEATURE_TORCH_MOBILE)
+namespace {
+  thread_local std::unique_ptr<c10::impl::IncludeTensorTypeIdGuard> guard_holder;
```

Comment:
There's no need to store an RAII struct in a unique pointer thread local. We can just add raw getter/setter API to `c10/core/impl/LocalTensorTypeSet.h`, if AMP's API absolutely requires it (otherwise, much better to use scoped access!)

Reply:
Yes, storing a persistent RAII thing kind of defeats the purpose of RAII.  I was trying to follow existing flow as much as possible.  I didn't feel bold enough to implement raw getters/setters myself.

I think the autocast dispatch state needs to be persistent in some way if the API uses a context manager.  If we alternatively decide (for example) that autocastability is a trait of Tensors, and propagates in the same way as requires_gradness, we might not need a global state like that.  API is still very much up in the air...

Reply:
OK, I'll add the getter/setter for you.

Reply:
https://github.com/pytorch/pytorch/pull/29592

Reply:
TensorTypeId-based getter/setter: https://github.com/pytorch/pytorch/pull/30113